File: PortfolioTracker.API/Domain/Entities/Transaction.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

namespace PortfolioTracker.API.Domain.Entities;

public enum TransactionType
{
    Buy,

    Sell,
}

public class Transaction
{
    public int Id { get; set; }

    public int AssetId { get; set; }

    public Asset Asset { get; set; } = null!;

    public decimal Quantity { get; set; }

    public decimal UnitPrice { get; set; }

    public decimal TotalAmount { get; set; }

    public string? Note { get; set; }

    public TransactionType Type { get; set; }

    [Column("TransactionDate")]
    [JsonIgnore]
    public DateTime Date { get; set; } = DateTime.UtcNow;

    // Keeps the API contract readable while the database column remains TransactionDate.
    [NotMapped]
    public DateTime TransactionDate
    {
        get => Date;
        set => Date = value;
    }
}

File: PortfolioTracker.Frontend/src/types/transaction.ts
export type TransactionType = "Buy" | "Sell";

export interface Transaction {
  id: number;
  symbol: string;
  quantity: number;
  unitPrice: number;
  totalAmount: number;
  note?: string | null;
  type: TransactionType;
  transactionDate: string;
}

export interface CreateTransactionDto {
  symbol: string;
  quantity: number;
  unitPrice: number;
  note?: string | null;
  type: TransactionType;
  transactionDate: string;
}

export interface UpdateTransactionDto extends CreateTransactionDto {
  id: number;
}

export interface TransactionQuery {
  type?: TransactionType;
  search?: string;
  symbol?: string;
}
